pkg/scm/downloaders/file/download.go

```diff
@@ -54,9 +55,27 @@ func (f *File) Download(config *api.Config) (*git.SourceInfo, error) {
 		return nil, lerr
 	}
 
+	var isIgnored func(path string) bool
+
+	if config.ExcludeRegExp != "" {
+		exclude, err := regexp.Compile(config.ExcludeRegExp)
+		if err != nil {
+			return nil, err
+		}
+		isIgnored = func(path string) bool {
+			_, ok := filesToIgnore[path]
+			return ok || exclude.MatchString(path)
+		}
+	} else {
+		isIgnored = func(path string) bool {
+			_, ok := filesToIgnore[path]
+			return ok
+		}
+	}
+
```

Comment:
how about something like this? (warning, untested)
```suggestion
isIgnored := func(path string) bool {
	if _, ok := filesToIgnore[path]; ok {
			  return ok
	}
	if len(config.ExcludeRegExp) != 0 {
			  return regexp.MustCompile(config.ExcludeRegExp).MatchString(path)
	}
}


```

Reply:
@coreydaley  That's quite a pesimisation. It compiles regexp for each check.

If I wanted to optimise code size I would do:
```go
	isIgnored := func(path string) bool {
		_, ok := filesToIgnore[path]
		return ok
	}

	if config.ExcludeRegExp != "" {
		exclude, err := regexp.Compile(config.ExcludeRegExp)
		if err != nil {
			return nil, err
		}
		old := isIgnored
		isIgnored = func(path string) bool {
			return old(path) || exclude.MatchString(path)
		}
	}
```
but that introduces one more indirect call via the closure.

Reply:
TBH  I am not huge fan of:
```go
filesToIgnore, lerr := di.GetListOfFilesToIgnore(copySrc)
```
It basically introduces one more traverse on top of the traverse done by `Copy()`.

IMHO the `ignore.DockerIgnorer` struct should have method `Matches(path string) bool` that could we used withing the `Copy()` traversal.

Reply:
Not that it made any visible difference on M.2 drives today.

Reply:
What I mean:
https://github.com/knative/func/blob/ac575e078f436df13839465d3fb31b2b97ab4a5c/pkg/pipelines/tekton/pipeplines_provider.go#L211-L219

there is even a library like that for .gitignore format.

Reply:
>It basically introduces one more traverse on top of the traverse done by `Copy()`.

Actually, maybe it's even worse, it traverse FS for each positive glob in `.s2iignore`.